app/models/label/LabelTable.scala

```diff
@@ -440,51 +448,78 @@ object LabelTable {
     var exists: Boolean = false
     var labelToValidate: List[(Int, String, String, Float, Float, Int, Int, Int, Int, Int)] = null
+    var selectedLabels: ListBuffer[Int] = labelIdList.getOrElse(new ListBuffer[Int]())
+
+    // TODO: add code that also checks that we havne't already chosen this label
+    val userIdString = userId.toString
+    val availableLabelCount: Int = getAvailableValidationLabels(userId, labelTypeId, labelIdList)
     while (!exists) {
-      val selectQuery = Q.query[Int, (Int, String, String, Float, Float, Int, Int, Int, Int, Int)](
-        """SELECT lb.label_id,
-        |       lt.label_type,
-        |       lb.gsv_panorama_id,
-        |       lp.heading,
-        |       lp.pitch,
-        |       lp.zoom,
-        |       lp.canvas_x,
-        |       lp.canvas_y,
-        |       lp.canvas_width,
-        |       lp.canvas_height
-        |FROM sidewalk.label AS lb,
-        |     sidewalk.label_type AS lt,
-        |     sidewalk.label_point AS lp,
-        |     sidewalk.gsv_data AS gd
-        |WHERE lp.label_id = lb.label_id
-        |      AND lt.label_type_id = lb.label_type_id
-        |      AND lb.label_type_id <> 5
-        |      AND lb.label_type_id <> 6
-        |      AND lb.deleted = false
-        |      AND lb.tutorial = false
-        |      AND gd.gsv_panorama_id = lb.gsv_panorama_id
-        |      AND gd.expired = false
-        |OFFSET floor(random() *
-        |      (
-        |          SELECT COUNT(*)
-        |          FROM sidewalk.label AS lb
-        |          WHERE lb.deleted = false
-        |              AND lb.tutorial = false
-        |              AND lb.label_type_id <> 5
-        |              AND lb.label_type_id <> 6
-        |      )
-        |)
-        |LIMIT ?""".stripMargin
-      )
-      val singleLabel = selectQuery(1).list
+      val r = new scala.util.Random
+      val labelOffset = r.nextInt(availableLabelCount - selectedLabels.length)
+
+      val labelsValidatedByUser = labelValidations.filter(_.userId === userIdString).map(_.labelId).list
+      var validationLabels = for {
+        _lb <- labels if _lb.labelTypeId === labelTypeId && _lb.deleted === false && _lb.tutorial === false
+        _lt <- labelTypes if _lt.labelTypeId === _lb.labelTypeId
+        _lp <- labelPoints if _lb.labelId === _lp.labelId
+        _gd <- gsvData if _gd.gsvPanoramaId === _lb.gsvPanoramaId && _gd.expired === false
+        _ms <- missions if _ms.missionId === _lb.missionId && _ms.userId =!= userIdString
+      } yield (_lb.labelId, _lt.labelType, _lb.gsvPanoramaId, _lp.heading, _lp.pitch, _lp.zoom,
+        _lp.canvasX, _lp.canvasY, _lp.canvasWidth, _lp.canvasHeight)
+
+      validationLabels = validationLabels.filterNot(_._1 inSet labelsValidatedByUser)
+      validationLabels = validationLabels.filterNot(_._1 inSet selectedLabels)
+
+      val singleLabel = validationLabels.drop(labelOffset).take(1).list
```

Comment:
Can you specify the type for `singleLabel`? Unless I'm missing something, I think this line should just be:
```
val singleLabel: LabelValidationMetadata =
    LabelValidationMetadata.tupled(validationLabels.drop(labelOffset).take(1).list.head)
```

Then over the next few lines, instead of doing `singleLabel(0)._3` you can just do `singleLabel.gsvPanoramaId`.

You would then also make `labelToValidate` of type `LabelValidationMetadata` as well.

Of course, let me know if you did it this way for a reason.